packages/xrpc-server/src/server.ts

```diff
@@ -200,6 +254,56 @@ export class Server {
       }
     }
   }
+
+  protected async addSubscription(
+    nsid: string,
+    def: LexXrpcSubscription,
+    config: XRPCStreamHandlerConfig,
+  ) {
+    const assertValidXrpcParams = (params: unknown) =>
+      this.lex.assertValidXrpcParams(nsid, params)
+    this.subscriptions.set(
+      nsid,
+      new XrpcStreamServer({
+        noServer: true,
+        handler: async function* (req) {
+          try {
+            const url = new URL(req.url || '', 'http://x')
+            // authenticate request
+            const auth = await config.auth?.({ req })
+            if (isHandlerError(auth)) {
+              throw XRPCError.fromError(auth)
+            }
```

Comment:
There's another world in which auth (and even param validation?) happens during the websocket upgrade (i.e. in `enableStreaming()`) rather than after it.  The upside would be that you never establish the websocket, so it's less resource intensive overall.  I didn't do this because you have to write some raw HTTP headers directly to the socket, and it felt a little funky— this ended-up being easier and more straightforward to propagate the error info.  Also not sure exactly how a 400 or 401 during the websocket upgrade looks to clients, if we switch I would want to check that out.  Another consideration, if we were ever to run this protocol over TCP then it would probably all happen over the connection as we're doing here.

Reply:
Leaving this as-is for now, as I'm not getting fantastic signals around either the correctness or client support for erroring during the websocket handshake.  It's also awkward to code-up using the ws module, as it requires writing some raw http.  I do also kind of like that these errors happen in-band, since it adapts naturally to TCP, etc.

Refs:
https://stackoverflow.com/questions/21762596/how-to-read-status-code-from-rejected-websocket-opening-handshake-with-javascrip/50685387#50685387
https://bugs.webkit.org/show_bug.cgi?id=80362